text/3512-closure-move-bindings.md

```diff
@@ -0,0 +1,284 @@
+- Feature `closure_move_bindings`
+- Start Date: 2023-10-09
+- RFC PR: [rust-lang/rfcs#3512](https://github.com/rust-lang/rfcs/pull/3512)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Adds the syntax `move(bindings) |...| ...`
+to explicitly specify how to capture bindings into a closure.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there are two ways to capture local bindings into a closure,
+namely by reference (`|| foo`) and by moving (`move || foo`).
+This mechanism has several ergonomic problems:
+
+- It is not possible to move some bindings and reference the others.
+To do so, one must define another binding that borrows the value
+and move it into the closure:
+
+```rs
+{
+    let foo = &foo;
+    move || run(foo, bar)
+}
+```
+
+- It is a very frequent scenario to clone a value into a closure
+(especially common with `Rc`/`Arc`-based values),
+but even the simplest scenario requires three lines of boilerplate:
+
+```rs
+{
+    let foo = foo.clone();
+    move || foo.run()
+}
+```
+
+This RFC proposes a more concise syntax to express these moving semantics.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A closure may capture bindings in its defining scope.
+Bindings are captured by reference by default:
```

Comment:
This is not accurate.
Bindings are "captured by example", so if the closure needs a reference to the value it's captured by reference and if needs an owned value it's captured by move...except for `Copy` types which always capture by reference unless you `move`.

Issue with more details:
https://github.com/rust-lang/rust/issues/100905

I think we should consider changing this behavior in a new edition and seeing if this helps without needing new syntax.
Personally for me every time I need the workarounds described in this RFC it was because of this.

Reply:
Does it sound reasonable to detect this as a breaking change and fix/warn in automated edition migration tools?